pkg/tkctl/cmd/check/check.go

```diff
@@ -0,0 +1,460 @@
+	var pdcfg *config_v2_1_16.PdConfig
+	var tikvcfg *config_v2_1_16.TikvConfig
+	if cfg, exist := tidbConfigMap.Data["config-file"]; exist {
+		infoPrintln("#### TiDB Config ####")
+		fmt.Println(cfg)
+		infoPrintln("#### TiDB Config ####\n")
+		_, err = toml.Decode(cfg, &tidbcfg)
+		if err != nil {
+			return fmt.Errorf("failed to decode tidb config: %v", err)
+		}
+		if tidbcfg.Log.Level != "info" {
+			warnPrintln("TiDB log-level:", tidbcfg.Log.Level)
+		}
+		if tidbcfg.Binlog.IgnoreError {
+			warnPrintln("TiDB binlog.ignore-error is enabled")
+		}
+		if tidbcfg.PreparedPlanCache.Enabled {
+			warnPrintln("TiDB prepared-plan-cache is experimental feature")
+		}
+		if tidbcfg.TxnLocalLatches.Enabled {
+			warnPrintln("TiDB txn-local-latches is for write conflict heavy business")
+		}
+	}
+	if cfg, exist := tikvConfigMap.Data["config-file"]; exist {
+		infoPrintln("#### TiKV Config ####")
+		fmt.Println(cfg)
+		infoPrintln("#### TiKV Config ####\n")
+		_, err = toml.Decode(cfg, &tikvcfg)
+		if err != nil {
+			return fmt.Errorf("failed to decode tikv config: %v", err)
+		}
+		tikvResources := util.ResourceRequirement(tidbcluster.Spec.TiKV.ContainerSpec)
+		memoryLimit := tikvResources.Limits[corev1.ResourceMemory]
+		m, b1 := memoryLimit.AsInt64()
+		if b1 {
+			// rocksdb.writecf.block-cache-size 10%~30% of memory limit
+			// rocksdb.defaultcf.block-cache-size 30%~50% of memory limit
+			if int64(tikvcfg.Rocksdb.Writecf.BlockCacheSize) < m/10 || int64(tikvcfg.Rocksdb.Writecf.BlockCacheSize) > 3*m/10 {
+				errorPrintf("TiKV rocksdb.writecf.block-cache-size should be 10%% ~ 30%% of memory limit, but got %d/%d\n", tikvcfg.Rocksdb.Writecf.BlockCacheSize, m)
+			}
+			if int64(tikvcfg.Rocksdb.Defaultcf.BlockCacheSize) < 3*m/10 || int64(tikvcfg.Rocksdb.Defaultcf.BlockCacheSize) > m/2 {
+				errorPrintf("TiKV rocksdb.writecf.block-cache-size should be 30%% ~ 50%% of memory limit, but got %d/%d\n", tikvcfg.Rocksdb.Defaultcf.BlockCacheSize, m)
+			}
+			infoPrintf("TiKV rocksdb.lockcf.block-cache-size: %dMiB\n", tikvcfg.Rocksdb.Lockcf.BlockCacheSize/1024/1024)
+		}
+		if !tikvcfg.RaftStore.SyncLog {
+			errorPrintln("Raftstore sync-log is disable")
+		}
+		if tikvcfg.LogLevel != "info" {
+			warnPrintln("TiKV log-level:", tikvcfg.LogLevel)
+		}
+	}
+	if cfg, exist := pumpConfigMap.Data["pump-config"]; exist {
+		infoPrintln("#### Pump Config ####")
+		fmt.Println(cfg)
+		infoPrintln("#### Pump Config ####\n")
+	}
+	if cfg, exist := pdConfigMap.Data["config-file"]; exist {
+		infoPrintln("#### PD Config ####")
+		fmt.Println(cfg)
+		infoPrintln("#### PD Config ####\n")
+		_, err = toml.Decode(cfg, &pdcfg)
+		if err != nil {
+			return fmt.Errorf("failed to decode pd config: %v", err)
+		}
+		if pdcfg.Log.Level != "info" {
+			warnPrintln("PD log-level:", pdcfg.Log.Level)
+		}
+		if pdcfg.Replication.MaxReplicas != 3 {
+			warnPrintln("TiKV data replication factor is", pdcfg.Replication.MaxReplicas)
+		}
+	}
+	return nil
```

Comment:
For better CI integration, I suggest this command exit with non-zero code upon any errors by default(some errors are just printed), and could be configured to also exit with non-zero code upon warnings.